src/metabase/db/data_migrations.clj

```diff
@@ -19,6 +26,30 @@
 
 (models/defmodel ^:deprecated DataMigrations :data_migrations)
 
+(setting/defsetting data-migration-index
+  "Used during data-migration to check if a data-migration is already ran before"
+  :type       :integer
+  :visibility :internal)
+
+(defn should-run-data-migration?
+  "`true` if the current `data-migration-index` is lower than the required index"
+  [required-data-migration-index]
+  (let [data-migration-index (setting/get :data-migration-index)]
+    (or (nil? data-migration-index) (< data-migration-index required-data-migration-index))))
+
+(defmacro run-with-data-migration-index
+  "To avoid re-running a data-migration, we use `:data-migration-index` setting to check whether or not a migration is already run.
+  Useful when we need to enforce a data-migration to only run once when upgrading Metabase.
+
+  When defining a new data migration, make sure the `data-migration-index` is incremented from the last data-migration,
+  otherwise the migration will be skipped even when upgrading Metabase.
+  After running migration successfully, the `data-migration-index` will be automatically updated"
+  {:style/indent 1}
+  [data-migration-index & body]
+  `(when (should-run-data-migration? ~data-migration-index)
+    ~@body
+    (setting/set-value-of-type! :integer :data-migration-index ~data-migration-index)))
```

Comment:
This is the main change that related to the introduction of `data-migration-index` setting.

The purpose of this is to introduce a mechanism so that we could write a data-migration that is guaranteed to run once when customers upgrade Metabase.

The idea is we store an index in the `setting` table, whenever we need a data migration to run once, we define that migration with an index.
And whenever data-migrations is triggered, it will skip all data migration with an index lower than the index stored in the `setting` table.